src/SplitTuples.cpp

```diff
@@ -335,10 +337,126 @@ class SplitTuples : public IRMutator {
     }
 };
 
+class SplitTupleExprs : public IRMutator {
+    using IRMutator::visit;
+
+    // TODO: worry about LetStmts that have tuple intrinsics in the RHS
+
+    Stmt visit(const Provide *op) override {
+        class GetTupleSize : public IRVisitor {
+            bool permitted = true;
+            using IRVisitor::visit;
+            void visit(const Call *op) override {
+                if (op->is_intrinsic(Call::tuple)) {
+                    user_assert(permitted)
+                        << "Can't nest an expression tuple inside another in definition of "
+                        << op->name << "\n";
+                    if (result == 0) {
+                        result = (int)op->args.size();
+                    } else {
+                        user_assert((int)op->args.size() == result)
+                            << "Expression tuples of mismatched sizes used in definition of "
+                            << op->name << ": " << result << " vs " << op->args.size();
+                    }
+                    // No nesting tuples
+                    permitted = false;
+                    IRVisitor::visit(op);
+                    permitted = true;
+                } else {
+                    IRVisitor::visit(op);
+                }
+            }
+
+        public:
+            int result = 0;
+        } get_tuple_size;
+
+        op->accept(&get_tuple_size);
+        int size = get_tuple_size.result;
+
+        if (size == 0) {
+            return IRMutator::visit(op);
+        }
+
+        // The LHS should contain at least one tuple, or our scatters
+        // all go to the same place. Is it worth asserting this? It
+        // could be a bug, or it could be some sort of degenerate base case.
+
+        // Fork the args and the RHS into their various versions
+        class ExtractTupleElement : public IRMutator {
+            using IRMutator::visit;
+            Expr visit(const Call *op) override {
+                if (op->is_intrinsic(Call::tuple)) {
+                    // No need to recursively mutate because we've
+                    // already asserted that these aren't nested.
+                    internal_assert(idx < (int)op->args.size());
+                    return op->args[idx];
+                } else {
+                    return IRMutator::visit(op);
+                }
+            }
+
+        public:
+            int idx;
+        } extractor;
+
+        vector<Stmt> provides;
+        vector<string> names;
+        vector<Expr> rhs_values;
+        for (extractor.idx = 0; extractor.idx < size; extractor.idx++) {
+            vector<Expr> args = op->args;
+            for (Expr &a : args) {
+                a = extractor.mutate(a);
+            }
+            vector<Expr> values = op->values;
```

Comment:
It looks like this is supporting `Tuple`s of `Call::tuple`s... should the test have one of these to exercise this?

Reply:
That's the intent but it doesn't work yet due to missing handling of LetStmts in the visitor

Reply:
Ok, pushed. I added a lexicographic bubble sort as another example. I don't think we could express bubble sort before.